core/packages/contracts/contracts/SovereignTreasury.sol

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.9;
+
+import "./Vault.sol";
+
+// TODO: transfer ownership from deployer to inbound channel
+contract SovereignTreasury is Ownable {
+    Vault public vault;
+
+    constructor(Vault _vault) {
+        vault = _vault;
+    }
+
+    // Handle a message from the bridge.
+    function handle(bytes32 _sovereignID, bytes calldata _message) external onlyOwner {
+        Message memory message = abi.decode(_message, (Message));
+
+        if (message.action == Action.Withdraw) {
+            // TODO: how do we know that the recipient address is payable?
+            WithdrawPayload memory payload = abi.decode(message.payload, (WithdrawPayload));
+
+            transfer(_sovereignID, payload.recipient, payload.amount);
+        }
+    }
+
+    // Deposit ETH into a sovereign account. Permissionless.
+    function deposit(bytes32 _sovereignID) external payable {
+        vault.deposit{ value: msg.value }(_sovereignID);
+    }
+
+    function transfer(bytes32 sovereignID, address payable recipient, uint256 amount) private {
+        vault.withdraw(sovereignID, recipient, amount);
+    }
+}
+
+struct Message {
+    Action action;
+    bytes payload;
+}
+
+enum Action {
+    // Withdraw from sovereign account and transfer to recipient.
+    // Parachain teams will occasionally send this message to retrieve collected fees.
+    Withdraw
+}
+
+struct WithdrawPayload {
+    address payable recipient;
```

Comment:
`address payable` is [not part of the ABI](https://docs.soliditylang.org/en/latest/abi-spec.html#mapping-solidity-to-abi-types), so here we're assuming that the recipient is payable.